src/box/memtx_hash.cc

```diff
@@ -602,6 +609,45 @@ hash_read_view_create_iterator(struct index_read_view *base,
 	return hash_read_view_iterator_start(it, type, key, part_count);
 }
 
+/** Implementation of dump_sort_data memtx_index_read_view callback. */
+static int
+hash_read_view_dump_primary_key(
+	struct memtx_index_read_view *base, ssize_t tuple_count,
+	struct memtx_sort_data *msd, bool *have_more)
+{
+	struct hash_read_view *rv = (struct hash_read_view *)base;
+
+	/* Collect the data to save. */
+	*have_more = true;
+	struct tuple **buffer =
+		(struct tuple **)xcalloc(tuple_count, sizeof(*buffer));
+	ssize_t dumped = 0;
+	while (dumped != tuple_count) {
+		struct tuple **ptuple = light_index_view_iterator_get_and_next(
```

Comment:
I don't think we can assume that tuples in a hash index will have the same order after recovery. I think we should be disable the feature if the primary index is hash.

Reply:
The tuples are recovered in the order they're written into the snapshot. And the order we write into the snapshot is strictly specified for a given hash index read view (we create an iterator to the first slot and step up to the last one, exactly as we dump the data into the sort data file). So the feature is safe for the HAS index.

Reply:
Right, thanks for the explanation. If both `.snap` and `.sortdata` were written in the same loop, it'd be easier to understand IMO.

Reply:
Done, see the [comment above](https://github.com/tarantool/tarantool/pull/11133#discussion_r2052029516).